man/disk161.1
.Dd January 5, 2015
.Dt DISK161 1
.Os System/161 2.x
.Sh NAME
.Nm disk161
.Nd System/161 disk utility
.Sh SYNOPSIS
.Nm disk161
create
.Op Fl f
.Ar filename
.Ar size
.Nm disk161
info
.Ar filename ...
.Nm disk161
resize
.Ar filename
.Ar delta-size
.Sh DESCRIPTION
The
.Nm disk161
utility manages System/161 disk images.
.Bl -tag -width createzz
.It Dv create
When run with the
.Dv create
command, it creates a new disk image of the requested size.
Unless the
.Fl f
(force) option is given,
.Nm disk161
will not erase or clobber an already-existing file.
.It Dv info
When run with the
.Dv info
command,
.Nm disk161
prints information about one or more disk image files.
.It Dv resize
When run with the
.Dv resize
command,
.Nm disk161
changes the size of an existing disk image file by the requested
amount
.Ar delta-size ,
which may be positive or negative.
Note that
.Nm disk161
changes the physical size of the disk; if the disk image contains a
file system or other data, one must also take file-system-level steps
to resize the logical structure.
In particular, shrinking a disk image without first shrinking the file
system on it will throw away data and often fatally corrupt the file
system.
.El
.Pp
The
.Ar size
and
.Ar delta-size
values may be given in bytes (plain number, or "b" suffix),
sectors ("s" suffix),
kilobytes ("K" suffix),
or (traditional) megabytes and gigabytes ("M" and "G" suffixes
respectively);
sizes must be a multiple of the sector size, which is 512 bytes.
.Pp
Note that System/161 disk images contain a 1-sector header with a
signature string; this is to help prevent accidents.
.Sh SEE ALSO
.Xr sys161 1
.Sh BUGS
System/161 only supports native System/161 disk images, which are
plain raw images with a 512-byte header.
It would be helpful to support compression and possibly other more
advanced formats, such as the
.Xr qemu 1
.Dv qcow2
format.


CHANGES
20160113 gwa		Add interval support for stat161.
20160108 dholland	Print why if disk geometry initializion fails.
20160105 dholland	Standardize and document the exit codes.
20160105 dholland	Allow K, M, etc. suffixes for disk/memory sizes.
20160105 dholland	Make the trace device reject (vs. ignore) invalid args.
20160105 dholland	Add -C option to supply device config arguments.

20151222 dholland	System/161 2.0.4 released.
20151222 dholland	Improve the way -X works.
20151222 dholland	Make trace161 -h output include the trace flag list.
20151222 dholland	Use shell printf instead of echo -n. Apparently Apple
........		still ships a broken sh whose echo prints literal "-n".
20151222 dholland	Apparently some Linux installs demand _GNU_SOURCE to
........		be allowed to see/use ftruncate.
20150808 dholland	Actually install the new prof.html. Sigh.
........		(issued as patch for 2.0.3)

20150804 dholland	System/161 2.0.3 released.
20150714 dholland	Fix the serial port FORCE logic meant to prevent
........		interrupt flapping. Thanks to Winnie Wu for a test
........		kernel and workload that reliably repeated the problem.
20150709 dholland	Recognize some obvious aliases for disk161 commands.
20150706 dholland	Add page on profiling to the manual.
20150706 dholland	Add dynamic profiling control to the trace device.
20150703 dholland	If profiling, write the profile out if we die().
20150703 dholland	Include seek times in disk activity trace output.

20150126 dholland	System/161 2.0.2 released.
20150126 dholland	Fix brown paper bag release. Sigh.
20150126 dholland	System/161 2.0.1 released.
20150121 dholland	Fix up some gdb threads issues appearing with gdb 7.8.
20150117 dholland	Update included .hgignore file, from Cary Gray.

20150115 dholland	System/161 2.0 released.
20150115 dholland	Add workaround to make disk rotdelay deterministic.
20150114 dholland	Fix stdint.h-related build problems on Linux.
20150114 dholland	Fix flock-related build problem on Linux.
20150109 dholland	Add software debugger requests to the trace device.
20150105 dholland	Make the testsuite run.
20150105 dholland	Add support for installing into a chroot/DESTDIR.
20150105 dholland	Add man pages for all executables.

20141209 dholland	System/161 1.99.10 released.
20141209 dholland	Add workaround for OpenBSD's srand/srandom breakage.

20140929 dholland	System/161 1.99.09 released.
20140929 dholland	Support the CACHE instruction as (for now) a nop.
20140929 dholland	Pretend to have 4K each L1 I/D caches.
20140929 dholland	Provide the MIPS32 config0 and config1 registers.
20140929 dholland	Edit/revise the processor docs.
20140929 dholland	Don't allow attempts to do r3k-style cache flushes.
20140929 dholland	Sort out status register definition going forward.
20140925 dholland	Fix address space ID printouts. From Peter Bailis.
20140925 dholland	Update disk config in sys161.conf.sample.
20140925 dholland	Added disk161 utility for manipulating disk images.
20140925 dholland	Warn if the size of a disk in sys161.conf is wrong.
20140925 dholland	Take the size of a disk from its image file size.
20140919 dholland	Fix build with gcc 4.8.
20140919 dholland	Provide flock compat for legacy OSes like Solaris.

20140821 dholland	System/161 1.99.08 released.
20140807 dholland	Use flock() on disk images to avoid accidents.
20140801 dholland	Improve gdb interface to treat cpus as "threads".
20140730 dholland	Rework tty handling. Now behaves when backgrounded.
20140730 dholland	Change sample disk image names to LHD{0,1}.conf.
20140730 dholland	Retune main loop now that timing works. Much faster.
20140730 dholland	Rework timing. Fix bogus large idle counts for real.
20140729 dholland	Rework main loop. Tidier and noticeably faster.
20140315 dholland	Handle EOF on input better, from Kevin Elphinstone.
20140314 dholland	Use -Wstrict-prototypes -Wmissing-prototypes with gcc.
20140123 dholland	Fix date on front page of docs.

20140123 dholland	System/161 1.99.07 released.
20140123 dholland	Add a -D option to enable a disk write doom counter.
20140122 dholland	Skip ELF note sections; from Vincent Danjean.
20131105 dholland	Change the processor ID value; discussion in docs.
20131105 dholland	Add mips coprocessor register banks ("selects").
20130530 dholland	Add a -X option to not hang waiting for the debugger.
20130530 dholland	Print more sensible cycle counts with stat161.
			(They are now totals across cpus, not averages.)
20130530 dholland	Add a -Z timeout option for monitoring user progress.
20130530 dholland	Count and report ll/sc and sync instructions.
20130530 dholland	Make configure script no longer leave "a.out" behind.
20130530 dholland	Purge ancient rcsid strings.
20130530 dholland	Accept the SYNC (memory barrier) instruction.
20130530 dholland	Have configure script check for _FILE_OFFSET_BITS=64.
20130422 dholland	Tidy up the disk initialization code further.
20130422 dholland	Fix missing initialization of dd_timedop in dev_disk.
			(Reported by Keno Fischer.)

20130422 dholland	System/161 1.99.06 released.
20130422 dholland	Update download URL link provided in the docs.
20130422 dholland	Fix one source of bogusly large idle cycle counts.
20130306 dholland	Hack around C99 strict-aliasing issues.
20130306 dholland	Fix build with gcc 4.5.
20130306 dholland	Initialize emufs create op's isdir return properly.
20130228 dholland	Doc fixes: typos, clarifications. Mention mips SYNC.

20110126 dholland	System/161 1.99.05 released.
20100819 dholland	Add a FORCE bit to the serial port device. See docs.

20090414 dholland	System/161 1.99.04 released.
20090414 dholland	Improve realism of disk timing model some.

20090228 dholland	System/161 1.99.03 released.
20090228 dholland	Fix time-handing bug that manifested under OS/161 2.x.
20090210 dholland	In stat161, fix reconnection-related printout glitches.
20090210 dholland	Improve interrupt/exception tracing.
20090210 dholland	Fix build issue on NetBSD-current.

20090206 dholland	System/161 1.99.02 released.

disktool/disk161.c
/*
 * disk161
 *
 * Disk image tool for sys161.
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>

/*
 * On Linux, or at least in some glibc versions, fcntl.h defines the
 * constants for flock(), but not the function, which is in sys/file.h
 * with a separate copy of the constants.
 */
#if defined(__linux__)
#include <sys/file.h>
#endif

#define SECTORSIZE 512
#define MINSIZE (128 * SECTORSIZE)
#define MAXSIZE 0x100000000LL

#define HEADERSIZE   SECTORSIZE
#define HEADERSTRING "System/161 Disk Image"

////////////////////////////////////////////////////////////
// Compat
/* XXX two copies of this are pasted here and in dev_disk.c */

/*
 * Apparently (as of September 2014) Solaris still doesn't have
 * flock(). What is this, 1990? Provide a wrapper around fcntl locks.
 *
 * If you don't have fcntl locks either, you're out of luck.
 */
#ifndef LOCK_EX

/*      LOCK_SH F_RDLCK - we don't use this */
#define LOCK_EX F_WRLCK
#define LOCK_UN F_UNLCK
#define LOCK_NB 0  /* assume we always want this */

#define flock(fd, op) myflock(fd, op)

static
int
myflock(int fd, int op)
{
	struct flock glop;

	glop.l_start = 0;
	glop.l_len = 0;
	glop.l_pid = getpid();
	glop.l_type = op;
	glop.l_whence = SEEK_SET;
	return fcntl(fd, F_SETLK, &glop);
}

#endif /* LOCK_EX */

////////////////////////////////////////////////////////////
// error wrappers

static
int
doopen(const char *file, int flags, mode_t mode)
{
	int fd;

	fd = open(file, flags, mode);
	if (fd < 0) {
		fprintf(stderr, "disk161: %s: %s\n",
			file, strerror(errno));
		exit(1);
	}

	return fd;
}

static
void
doread(const char *file, int fd, void *buf, size_t len)
{
	ssize_t r;

	r = read(fd, buf, len);
	if (r < 0) {
		fprintf(stderr, "disk161: %s: read: %s\n", file,
			strerror(errno));
		exit(1);
	}
	if ((size_t)r != len) {
		fprintf(stderr, "disk161: %s: reaD: Unexpected EOF\n", file);
		exit(1);
	}
}

static
void
dowrite(const char *file, int fd, const void *buf, size_t len)
{
	ssize_t r;

	r = write(fd, buf, len);
	if (r < 0) {
		fprintf(stderr, "disk161: %s: write: %s\n", file,
			strerror(errno));
		exit(1);
	}
	if ((size_t)r != len) {
		fprintf(stderr, "disk161: %s: write: Unexpected short count "
			"%zd of %zu\n", file, r, len);
		exit(1);
	}
}

static
void
dolseek(const char *file, int fd, off_t pos, int whence)
{
	if (lseek(fd, pos, whence) == -1) {
		fprintf(stderr, "disk161: %s: lseek: %s\n", file,
			strerror(errno));
	}
}

static
void
dotruncate(const char *file, int fd, off_t size)
{
	if (ftruncate(fd, size) < 0) {
		fprintf(stderr, "disk161: %s: %s\n", file, strerror(errno));
		exit(1);
	}
}

static
void
dofstat(const char *file, int fd, struct stat *st)
{
	if (fstat(fd, st) == -1) {
		fprintf(stderr, "disk161: %s: fstat: %s\n", file,
			strerror(errno));
		exit(1);
	}
}

static
void
doflock(const char *file, int fd, int mode)
{
	if (flock(fd, mode|LOCK_NB) == -1) {
		if (errno == EAGAIN) {
			fprintf(stderr, "disk161: %s: "
				"Locked by another process\n", file);
		}
		else {
			fprintf(stderr, "disk161: %s: flock: %s\n", file,
				strerror(errno));
		}
		exit(1);
	}
}

////////////////////////////////////////////////////////////
// common

static
off_t
getsize(const char *str)
{
	char *suffix;
	off_t value;

	errno = 0;
	value = strtoll(str, &suffix, 0);
	if (errno) {
		fprintf(stderr, "disk161: %s: Invalid number\n", str);
		exit(1);
	}
	if (*suffix == 0 || !strcmp(suffix, "b")) {
		/* bytes */
		return value;
	}
	if (!strcmp(suffix, "s")) {
		/* sectors */
		return value * SECTORSIZE;
	}
	if (!strcmp(suffix, "k") || !strcmp(suffix, "K")) {
		/* Kb */
		return value*1024;
	}
	if (!strcmp(suffix, "m") || !strcmp(suffix, "M")) {
		/* Mb */
		return value*1024*1024;
	}
	if (!strcmp(suffix, "g") || !strcmp(suffix, "G")) {
		/* Gb */
		return value*1024*1024*1024;
	}
	fprintf(stderr, "disk161: Invalid size suffix %s\n", suffix);
	exit(1);
}

static
off_t
filesize(const char *file, int fd)
{
	struct stat st;

	dofstat(file, fd, &st);
	return st.st_size;
}

static
void
checksize(off_t size)
{
	if (size % SECTORSIZE) {
		fprintf(stderr, "Size %lld not an even number of sectors\n",
			(long long)size);
		size = SECTORSIZE * ((size + SECTORSIZE - 1) / SECTORSIZE);
		fprintf(stderr, "Try %lld instead.\n", (long long)size);
		exit(1);
	}
	if (size < MINSIZE) {
		fprintf(stderr, "Size %lld too small\n", (long long)size);
		exit(1);
	}
	if (size >= MAXSIZE) {
		fprintf(stderr, "Size %lld too large\n", (long long)size);
		exit(1);
	}
}

static
void
checkheader(const char *file, int fd)
{
	char buf[SECTORSIZE];

	doread(file, fd, buf, sizeof(buf));
	buf[sizeof(buf) - 1] = 0;

	if (!strcmp(buf, HEADERSTRING)) {
		/* ok */
		return;
	}
	fprintf(stderr, "disk161: %s: Not a System/161 disk image\n", file);
	exit(1);
}

static
void
writeheader(const char *file, int fd)
{
	char buf[SECTORSIZE];

	memset(buf, 0, sizeof(buf));
	strcpy(buf, HEADERSTRING);

	dolseek(file, fd, 0, SEEK_SET);
	dowrite(file, fd, buf, sizeof(buf));
}

////////////////////////////////////////////////////////////
// create

static
void
docreate(const char *file, const char *sizespec, int doforce)
{
	int fd;
	off_t size;

	if (!doforce) {
		fd = open(file, O_RDONLY);
		if (fd >= 0) {
			fprintf(stderr, "disk161: %s: %s\n", file,
				strerror(EEXIST));
			exit(1);
		}
	}

	fd = doopen(file, O_RDWR|O_CREAT|O_TRUNC, 0664);
	doflock(file, fd, LOCK_EX);
	size = getsize(sizespec);
	checksize(size);
	dotruncate(file, fd, HEADERSIZE + size);
	writeheader(file, fd);
	doflock(file, fd, LOCK_UN);
	close(fd);
}

////////////////////////////////////////////////////////////
// info

static
void
doinfo(const char *file)
{
	int fd;
	struct stat st;
	long long amt;

	fd = doopen(file, O_RDWR, 0);
	checkheader(file, fd);
	dofstat(file, fd, &st);

	amt = st.st_size - HEADERSIZE;
	printf("%s size %lld bytes (%lld sectors; %lldK; %lldM)\n", file,
	       amt, amt / SECTORSIZE, amt / 1024, amt / (1024*1024));
	

	amt = st.st_blocks * 512LL;
	printf("%s spaceused %lld bytes (%lld sectors; %lldK; %lldM)\n", file,
	       amt, amt / SECTORSIZE, amt / 1024, amt / (1024*1024));

	close(fd);
}

////////////////////////////////////////////////////////////
// resize

static
void
doresize(const char *file, const char *sizespec)
{
	enum { M_SET, M_PLUS, M_MINUS } mode;
	off_t oldsize, newsize;
	int fd;

	if (*sizespec == '+') {
		sizespec++;
		mode = M_PLUS;
	}
	else if (*sizespec == '-') {
		sizespec++;
		mode = M_MINUS;
	}
	else {
		mode = M_SET;
	}

	newsize = getsize(sizespec);
	fd = doopen(file, O_RDWR, 0);
	doflock(file, fd, LOCK_EX);
	checkheader(file, fd);
	oldsize = filesize(file, fd);
	oldsize -= HEADERSIZE;
	switch (mode) {
	    case M_SET:
		break;
	    case M_PLUS:
		if (oldsize + newsize < oldsize) {
			/* overflow */
			fprintf(stderr, "+%s: Result too large\n", sizespec);
			exit(1);
		}
		newsize = oldsize + newsize;
		break;
	    case M_MINUS:
		if (oldsize < newsize) {
			/* underflow */
			fprintf(stderr, "-%s: Result too small\n", sizespec);
			exit(1);
		}
		newsize = oldsize - newsize;
		break;
	}

	checksize(newsize);
	dotruncate(file, fd, HEADERSIZE + newsize);
	doflock(file, fd, LOCK_UN);
	close(fd);
}

////////////////////////////////////////////////////////////
// main

static
void
usage(void)
{
	fprintf(stderr, "Usage: disk161 action [options] [arguments]\n");
	fprintf(stderr, "   disk161 create [-f] filename size\n"); 
	fprintf(stderr, "   disk161 info filename...\n");
	fprintf(stderr, "   disk161 resize filename [+-]size\n");
	exit(3);
}

int
main(int argc, char *argv[])
{
	const char *command;
	int doforce = 0;
	int ch;
	int i;

	if (argc < 2) {
		usage();
	}

	command = argv[1];
	argv++;
	argc--;

	while ((ch = getopt(argc, argv, "f"))!=-1) {
		switch (ch) {
		    case 'f': doforce = 1; break;
		    default: usage();
		}
	}

	if (!strcmp(command, "create")) {
		if (optind + 2 != argc) {
			usage();
		}
		docreate(argv[optind], argv[optind+1], doforce);
	}
	else if (!strcmp(command, "info") || !strcmp(command, "stat") ||
		 !strcmp(command, "stats") || !strcmp(command, "status")) {
		if (doforce) {
			usage();
		}
		for (i=optind; i<argc; i++) {
			doinfo(argv[i]);
		}
	}
	else if (!strcmp(command, "resize") || !strcmp(command, "setsize")) {
		if (optind + 2 != argc) {
			usage();
		}
		if (doforce) {
			usage();
		}
		doresize(argv[optind], argv[optind+1]);
	}
	else if (!strcmp(command, "help")) {
		usage();
	}
	else {
		fprintf(stderr, "disk161: Unknown command %s\n", command);
		usage();
	}

	return 0;
}

